### PR TITLE
Add full node snapshotting with enableFullNodeSnapshots flag

### PR DESCRIPTION
**Pull request by platfowner**
Summary:
- Add full node snapshotting with enableFullNodeSnapshots flag (fromSnapshotObject(), toSnapshotObject()) 
- Add prefix to both state label ('#state:') and radix label ('#radix:') in state proof

Related issues:
- https://github.com/ainblockchain/ain-blockchain/issues/649

### CODE REVIEW

File: db/radix-node.js [hunk 1]
```diff
@@ -546,13 +547,89 @@ class RadixNode {
     return numAffectedNodes;
   }
 
+  static parseChildStateNode(obj) {
```
Comment (liayoo) [inline db/radix-node.js:550]: how about parseChildStateNodeFromSnapshot() ?

File: db/index.js [hunk 2]
```diff
@@ -157,6 +164,18 @@ class DB {
     return true;
   }
 
+  /**
+   * Replaces the state root.
+   *
+   * @param {StateNode} newRoot new root to replace with
+   */
+  replaceStateRoot(newRoot) {
+    const newVersion = newRoot.getVersion();
```
Comment (liayoo) [inline db/index.js:173]: question) would there be a need to check if the newVersion doesn't have a conflicting version?

File: common/constants.js [hunk 3]
```diff
@@ -34,6 +34,8 @@ const FeatureFlags = {
   enableTrafficMonitoring: true,
   // Enables state info updates.
   enableStateInfoUpdates: true,
+  // Enables full node snapshots.
+  enableFullNodeSnapshots: true,
```
Comment (liayoo) [inline common/constants.js:38]: how about enableFullDbSnapshots or enableDetailedSnapshots?

File: db/index.js (hunk 2 above)
Comment (platfowner) [reply db/index.js:174]: Just added the conflict check. Thanks!

File: db/radix-node.js (hunk 1 above)
Comment (platfowner) [reply db/radix-node.js]: Changed to getChildStateNodeFromSnapshotObject(). PTAL

File: common/constants.js (hunk 3 above)
Comment (platfowner) [reply common/constants.js]: Changed to enableRadixLevelSnapshots. PTAL

**Comment by platfowner** [comment]
Comments addressed, PTAL! @liayoo

File: db/index.js [hunk 4]
```diff
@@ -157,6 +164,23 @@ class DB {
...
+   * Replaces the state root.
+   *
+   * @param {StateNode} newRoot new root to replace with
+   */
+  replaceStateRoot(newRoot) {
+    const LOG_HEADER = 'replaceStateRoot';
+    const newVersion = newRoot.getVersion();
+    if (this.stateManager.hasVersion(newVersion)) {
+      CommonUtil.finishWithStackTrace(
+          logger, `[${LOG_HEADER}] State version already exists: ${newVersion}`);
```
Comment (liayoo) [inline db/index.js:177]: should we process.exit() here?

**Review by liayoo — APPROVED** [review]
LGTM with a comment. Thanks for handling this!

File: db/index.js (hunk 4 above)
Comment (platfowner) [reply db/index.js:177]: Actually, finishWithStackTrace() tries to do similar with the some technical difficulties:
- If we call process.exit() directly somewhere in the code, the corresponding critical error messages might be skipped with the current winston logger's structure.
- So, we're just trying to stop logging after the critical log messages.

**Comment by platfowner** [comment]
Thanks for the review! Now merging..

File: db/index.js (hunk 4 above)
Comment (liayoo) [reply db/index.js:177]: Ah, right right. I was thinking of the old convention. Will keep that in mind, thanks!

**Pull request merged**